ports/nrf/supervisor/port.c

```diff
@@ -247,7 +248,15 @@ void port_sleep_until_interrupt(void) {
         sd_app_evt_wait();
     } else {
         // Call wait for interrupt ourselves if the SD isn't enabled.
+        // Note that `wfi` should be called with interrupts disabled,
+        // to ensure that the queue is properly drained.  The `wfi`
+        // instruction will returned as long as an interrupt is
+        // available, even though the actual handler won't fire until
+        // we re-enable interrupts.
+        common_hal_mcu_disable_interrupts();
```

Comment:
Yeah, I did a bit of research it is advisable prat to disable the interrupt before WFI() to avoid race condition. WFI() doesn't require interrupt enabled to wake CPU. It is best practice to also all DSB() before WFI() as well according to ARM docs. Here is the implementation of zephyr

https://github.com/zephyrproject-rtos/zephyr/blob/master/arch/arm/core/aarch32/cpu_idle.S#L75

There is a other useful doc from mbed as well.
https://os.mbed.com/media/uploads/pateshian/wfi_wake_up_from_sleep_11_17_2016_.pdf (D_SLEEP_VERSION == 2)

Reply:
another reference from freeRTOS tickless mode, which do the same as this function
https://github.com/FreeRTOS/FreeRTOS-Kernel/blob/master/portable/GCC/ARM_CM4F/port.c#L510

Reply:
I did experiment with `WFE` instead of `WFI`, however that resulted in the system never sleeping for very long.

In FreeRTOS, vPortSuppressTicksAndSleep() looks like needs to be called from a critical section (i.e. when interrupts are disabled): https://github.com/FreeRTOS/FreeRTOS-Kernel/blob/5d28744febb6788ff37ad51990848c7bd82296e6/include/task.h#L2538-L2544

Reply:
I'm sorry, I misunderstood the wording there.

`eTaskConfirmSleepModeStatus()` is called from `vPortSuppressTicksAndSleep()` when interrupts are disabled to determine if it's safe to perform the sleep. You could do something similar here, where you disable interrupts and then go through each subsystem to determine if there are any pending buffers to flush.

The approach taken in this patch is to just go ahead and flush the buffers.

Reply:
@xobs I just reference to freeRTOS to just point out that interrupt disabled is needed before entering WFI() as your PR does. The freeRTOS only enter sleep mode in idle task which guarantee there is no active event queue that is not processed. So yes, draining the event queue is a must, or we can just check the event queue, and skip the WFI until next time as suggested below which I think is a good approach as well.